File: src/com/prey/actions/location/PreyGooglePlayServiceLocation.java
package com.prey.actions.location;

import android.content.Context;
import android.location.Location;
import android.os.Bundle;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GooglePlayServicesClient;
import com.google.android.gms.location.LocationClient;
import com.google.android.gms.location.LocationListener;
import com.google.android.gms.location.LocationRequest;
import com.prey.PreyLogger;

public class PreyGooglePlayServiceLocation implements LocationListener, GooglePlayServicesClient.ConnectionCallbacks, GooglePlayServicesClient.OnConnectionFailedListener {
 
	/*
	 * Constants for location update parameters
	 */
	// Milliseconds per second
	public static final int MILLISECONDS_PER_SECOND = 1000;
	// The update interval
	public static final int UPDATE_INTERVAL_IN_SECONDS = 10;
	// A fast interval ceiling
	public static final int FAST_CEILING_IN_SECONDS = 1;
	// Update interval in milliseconds
	public static final long UPDATE_INTERVAL_IN_MILLISECONDS = MILLISECONDS_PER_SECOND * UPDATE_INTERVAL_IN_SECONDS;
	// A fast ceiling of update intervals, used when the app is visible
	public static final long FAST_INTERVAL_CEILING_IN_MILLISECONDS = MILLISECONDS_PER_SECOND * FAST_CEILING_IN_SECONDS;
	private LocationRequest mLocationRequest;
	private LocationClient mLocationClient;
	private Location currentLocation = null;

	public void init(Context ctx) {
		PreyLogger.d("init");
		// Create a new global location parameters object
		mLocationRequest = LocationRequest.create();
		/*
		 * Set the update interval
		 */
		mLocationRequest.setInterval(UPDATE_INTERVAL_IN_MILLISECONDS);
		// Use high accuracy
		mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
		// Set the interval ceiling to one minute
		mLocationRequest.setFastestInterval(FAST_INTERVAL_CEILING_IN_MILLISECONDS);
		mLocationClient = new LocationClient(ctx, this, this);
		mLocationClient.connect();
	}

	public Location getLastLocation(Context ctx) {
		try {
			if (currentLocation == null) {
				currentLocation = mLocationClient.getLastLocation();
			}
		} catch (Exception e) {
		}
		PreyLogger.d("getLastLocation is null:" + (currentLocation == null));
		return currentLocation;
	}

	/*
	 * Called by Location Services if the attempt to Location Services fails.
	 */
	@Override
	public void onConnectionFailed(ConnectionResult connectionResult) {
		PreyLogger.d("onConnectionFailed");
		/*
		 * Google Play services can resolve some errors it detects. If the error
		 * has a resolution, try sending an Intent to start a Google Play
		 * services activity that can resolve error.
		 */
		if (connectionResult.hasResolution()) {
			try {
				PreyLogger.d("CONNECTION_FAILURE_RESOLUTION_REQUEST");
				/*
				 * Thrown if Google Play services canceled the original
				 * PendingIntent
				 */
			} catch (Exception e) {
				PreyLogger.d("error:" + e.getMessage());
			}
		} else {
			PreyLogger.d("error:" + connectionResult.getErrorCode());
		}
	}

	@Override
	public void onConnected(Bundle arg0) {
		PreyLogger.d("onConnected");
		// TODO Auto-generated method stub
		startPeriodicUpdates();
	}

	@Override
	public void onDisconnected() {
		PreyLogger.d("onDisconnected");
		// TODO Auto-generated method stub
		stopPeriodicUpdates();
	}

	/**
	 * In response to a request to start updates, send a request to Location
	 * Services
	 */
	public void startPeriodicUpdates() {
		PreyLogger.d("startPeriodicUpdates");
		mLocationClient.requestLocationUpdates(mLocationRequest, this);
	}

	/**
	 * In response to a request to stop updates, send a request to Location
	 * Services
	 */
	public void stopPeriodicUpdates() {
		PreyLogger.d("stopPeriodicUpdates");
		mLocationClient.removeLocationUpdates(this);
	}

	@Override
	public void onLocationChanged(Location location) {
		PreyLogger.d("onLocationChanged is null" + (location == null));
		// In the UI, set the latitude and longitude to the value received
		// mLatLng.setText(LocationUtils.getLatLng(this, location));
		this.currentLocation = location;
	}
}

File: app/src/main/java/com/prey/actions/location/kotlin/PreyGooglePlayServiceLocation.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.actions.location.kotlin

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.os.Build
import android.os.Bundle
import android.os.Looper
import androidx.core.app.ActivityCompat
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.LocationListener
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import com.prey.kotlin.PreyConfig
import com.prey.kotlin.PreyLogger
import java.text.DateFormat
import java.util.Date
import java.util.concurrent.TimeUnit

class PreyGooglePlayServiceLocation : GoogleApiClient.ConnectionCallbacks,
    GoogleApiClient.OnConnectionFailedListener,
    LocationListener {
    protected var mGoogleApiClient: GoogleApiClient? = null
    protected var mLocationRequest: LocationRequest? = null
    protected var mCurrentLocation: Location? = null
    protected var mLastUpdateTime: String? = null
    protected var mRequestingLocationUpdates: Boolean? = null
    private var ctx: Context? = null

    fun init(ctx: Context?) {
        this.ctx = ctx
        mCurrentLocation = null
        mLastUpdateTime = null
        mRequestingLocationUpdates = false
        buildGoogleApiClient()
    }

    fun getLastLocation(ctx: Context?): Location? {
        return mCurrentLocation
    }

    @Synchronized
    protected fun buildGoogleApiClient() {
        mGoogleApiClient = GoogleApiClient.Builder(ctx!!)
            .addConnectionCallbacks(this)
            .addOnConnectionFailedListener(this)
            .addApi(LocationServices.API)
            .build()
        val connectionResult = mGoogleApiClient!!.blockingConnect(
            GOOGLE_API_CLIENT_TIMEOUT_S.toLong(), TimeUnit.SECONDS
        )
        if (connectionResult.isSuccess && mGoogleApiClient!!.isConnected) {
            createLocationRequest()
            startLocationUpdates()
        } else {
            PreyLogger.d(
                String.format(
                    GOOGLE_API_CLIENT_ERROR_MSG,
                    connectionResult.errorCode
                )
            )
        }
    }

    protected fun createLocationRequest() {
        mLocationRequest = LocationRequest()
        mLocationRequest!!.setInterval(PreyConfig.UPDATE_INTERVAL_IN_MILLISECONDS)
        mLocationRequest!!.setFastestInterval(PreyConfig.FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS)
        mLocationRequest!!.setPriority(PreyConfig.LOCATION_PRIORITY_HIGHT)
    }

    override fun onConnected(connectionHint: Bundle?) {
    }

    override fun onLocationChanged(location: Location) {
        mCurrentLocation = location
        if (location != null) {
            mLastUpdateTime = DateFormat.getTimeInstance().format(Date())
            stopLocationUpdates()
        }
    }

    fun stopLocationUpdates() {
        LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient!!, this)
    }

    override fun onConnectionSuspended(cause: Int) {
        PreyLogger.d("Connection suspended")
        mGoogleApiClient!!.connect()
    }

    override fun onConnectionFailed(result: ConnectionResult) {
        PreyLogger.d("Connection failed: ConnectionResult.getErrorCode() = " + result.errorCode)
    }

    protected fun startLocationUpdates() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M
            || (ActivityCompat.checkSelfPermission(
                ctx!!,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
                    && ActivityCompat.checkSelfPermission(
                ctx!!,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED)
        ) {
            try {
                Looper.prepare()
                LocationServices.FusedLocationApi.requestLocationUpdates(
                    mGoogleApiClient!!, mLocationRequest!!, this
                )
                Looper.loop()
            } catch (e: Exception) {
                PreyLogger.d("Error startLocationUpdates: " + e.message)
            }
        }
    }

    companion object {
        const val GOOGLE_API_CLIENT_TIMEOUT_S: Int = 10
        const val GOOGLE_API_CLIENT_ERROR_MSG: String =
            "Failed to connect to GoogleApiClient (error code = %d)"
    }
}